### ISSUE TITLE
Testing pymol could be easier and more robust

### ISSUE THREAD

**Issue by llimeht** [schrodinger/pymol-open-source#10]
Tests are important: users want to be able to test that their compilation worked OK, developers want to be able to use tests as part of CI, packagers want to be able to run tests on the packaged version to make sure that everything has indeed been packaged properly. 

This git repo contains a `test` directory which exercises some parts of `pymol`. Most of these tests currently fail because pymol has changed its output format over the years:

```
 run_tests: C0001quit...
 run_tests: C0010hello...
 run_tests: C0020pdb...
 run_tests: C0020pdb.log DIFFERS over about 2 lines.
 run_tests: C0021pdbstr...
 run_tests: C0025mol...
 run_tests: C0025mol.log DIFFERS over about 2 lines.
 run_tests: C0026sdf...
 run_tests: C0026sdf.log DIFFERS over about 5 lines.
 run_tests: C0027mol2...
 run_tests: C0027mol2.log DIFFERS over about 9 lines.
 run_tests: C0030pkl...
 run_tests: C0030pkl.pdb DIFFERS over about 196 lines.
 run_tests: C0050multi...
 run_tests: C0050multi.log DIFFERS over about 6 lines.
 run_tests: C0060delete...
 run_tests: C0060delete.log DIFFERS over about 6 lines.
 run_tests: C0070rep...
 run_tests: C0070rep.log DIFFERS over about 2 lines.
 run_tests: C0080create...
 run_tests: C0080create.log DIFFERS over about 6 lines.
 run_tests: C0100select...
 run_tests: C0100select.log DIFFERS over about 16 lines.
 run_tests: C0105names...
 run_tests: C0105names.log DIFFERS over about 2 lines.
 run_tests: C0106order...
 run_tests: C0106order.log DIFFERS over about 6 lines.
 run_tests: C0140iterate...
 run_tests: C0140iterate.log DIFFERS over about 12 lines.
 run_tests: C0150alter...
 run_tests: C0150alter.log DIFFERS over about 2 lines.
 run_tests: C0150alter.1.pdb DIFFERS over about 39 lines.
 run_tests: C0160get...
 run_tests: C0160get.log DIFFERS over about 6 lines.
 run_tests: C0170scene...
 run_tests: C0170scene.log DIFFERS over about 78 lines.
 run_tests: C0180embed...
 run_tests: C0180embed.log DIFFERS over about 11 lines.
 run_tests: C0181embed...
 run_tests: C0181embed.log DIFFERS over about 2 lines.
 run_tests: C0200feedback...
 run_tests: C0200feedback.log DIFFERS over about 40 lines.
 run_tests: C0201extend...
 run_tests: C0202help_python...
 run_tests: C0202help_python.log DIFFERS over about 18 lines.
 run_tests: C0300session...
 run_tests: C0300session.log DIFFERS over about 4 lines.
 run_tests: C0500misc...
 run_tests: C0500misc.log DIFFERS over about 33 lines.
 run_tests: C0550pseudo...
 run_tests: C0550pseudo.log DIFFERS over about 32 lines.
 run_tests: C0600util...
 run_tests: C0600util.log DIFFERS over about 17 lines.
 run_tests: C0700odd...
 run_tests: C0700odd.2.pdb DIFFERS over about 1 lines.
 run_tests: C0700odd.E.pdb DIFFERS over about 35 lines.
 run_tests: C0700odd.log DIFFERS over about 65 lines.
 run_tests: C0700odd.C.pdb DIFFERS over about 35 lines.
 run_tests: C0700odd.6.pdb DIFFERS over about 1 lines.
 run_tests: C0700odd.9.pdb DIFFERS over about 22 lines.
 run_tests: C0700odd.1.pdb DIFFERS over about 22 lines.
 run_tests: C0700odd.8.pdb DIFFERS over about 22 lines.
 run_tests: C0700odd.3.pdb DIFFERS over about 1 lines.
 run_tests: C0700odd.B.pdb DIFFERS over about 22 lines.
 run_tests: C0700odd.5.pdb DIFFERS over about 1 lines.
 run_tests: C0700odd.A.pdb DIFFERS over about 22 lines.
 run_tests: C0700odd.F.pdb DIFFERS over about 1 lines.
 run_tests: C0700odd.4.pdb DIFFERS over about 1 lines.
 run_tests: C0700odd.D.pdb DIFFERS over about 35 lines.
 run_tests: C0700odd.7.pdb DIFFERS over about 22 lines.
 run_tests: C0740multipdb...
 run_tests: C0800matrices...
 run_tests: C0800matrices.log DIFFERS over about 21 lines.
 run_tests: C1000viewing...
 run_tests: C1000viewing.log DIFFERS over about 124 lines.
 run_tests: C1010helping...
 run_tests: C1010helping.log DIFFERS over about 12 lines.
 run_tests: C1020querying...
 run_tests: C1020querying.log DIFFERS over about 27 lines.
 run_tests: C1030selecting...
 run_tests: C1030selecting.log DIFFERS over about 24 lines.
 run_tests: C1035wildcards...
 run_tests: C1035wildcards.log DIFFERS over about 2 lines.
 run_tests: C1040editing...
 run_tests: C1040editing.log DIFFERS over about 77 lines.
 run_tests: C1050fitting...
 run_tests: C1060controlling...
 run_tests: C1070exporting...
 run_tests: C1070exporting.log DIFFERS over about 11 lines.
 run_tests: C1080isomore...
 run_tests: C1080isomore.log DIFFERS over about 10 lines.
 run_tests: C1090coulomb...
 run_tests: C1090coulomb.log DIFFERS over about 2 lines.
 run_tests: C1095slice...
 run_tests: C1095slice.log DIFFERS over about 27 lines.
 run_tests: C1100do...
 run_tests: C1100do.log DIFFERS over about 14 lines.
 run_tests: C1110matrix_mode...
 run_tests: C1110matrix_mode.log DIFFERS over about 12 lines.
 run_tests: C1200matrix_maps...
 run_tests: C1200matrix_maps.log DIFFERS over about 14 lines.
 run_tests: C1300extras...
 run_tests: C2000returns...
 run_tests: C2000returns.log DIFFERS over about 126 lines.
```
with most output being reasonably trivial:
`diffs/C0020pdb.log.diff `
```
2c2
<  CmdLoad: "dat/pept.pdb" loaded as "pept".
---
>  CmdLoad: PDB-string loaded into object "pept", state 1.
```

`diffs/C1090coulomb.log.diff`
```
45c45
<  IsosurfVolume: Surface generated using 3351 lines.
---
>  IsosurfVolume: Surface generated using 5068 lines.
```
(the test runner also exits without error even though the tests failed)

There is additionally the [pymol-testing](https://github.com/schrodinger/pymol-testing) repo which contains a more extensive set of tests.

From my perspective as a packager who is also likely to send the occasional patch to you as we spot integration problems with newer versions of libraries etc, I'd love to see a more robust test suite within the pymol repository that I can use at packaging time as [CI within the distribution](http://ci.debian.net/packages/p/pymol) and on [GitHub](https://docs.travis-ci.com/user/tutorial/) when feeding patches back to you.

How can I help achieve that?

  * Do the tests in `test` still have value? Would you like a PR to update them so they all pass again?

  * Could you be convinced to import the [pymol-testing](https://github.com/schrodinger/pymol-testing) repo into this one as a replacement for the existing tests?

  * Would you be interested in hooking these into travis-ci so that each push and each PR got tested against Python 2.7, Python 3.x (for some value of x) and was shown to not break things in the way I accidentally did the other day in #9?

The current approach to testing left Python 3 errors unfixed even though Python 3 has been supported for over a year. The problems fixed in #9 were found only because we've been faking a test suite using the contents of `examples` and in switching the packaging to Python 3, the output changed. Shall we see if we can catch such problems automatically?

**Comment by JarrettSJohnson** [comment]
Thanks for opening a ticket regarding this matter and we apologize for the late response. First, I agree that tests are integral to the health and maintenance of a software product. For our day-to-day development, we have extensive, internal testing infrastructure plus QA personnel that assists in making sure PyMOL is as stable as possible. For the past couple of months, we've been discussing how to integrate some of the software management strategies to the open-source project and we're starting to have a clearer idea of what it would look like, many ideas that you suggested are things we are actually already considering including CI support and possible integration of [pymol-testing](https://github.com/schrodinger/pymol-testing) into the main development repo. Logistically, implementing these ideas is one of the goals for incoming PyMOL Fellows. 

I also want to point out that [pymol-open-source/test](https://github.com/schrodinger/pymol-open-source/tree/master/test) is no longer supported and will be practically replaced by [pymol-testing](https://github.com/schrodinger/pymol-testing) and the internal C++ tests. Internally, we test against both Python 2.7 and Python 3.6, but perhaps for the future, we may need to consider streamlining Github PRs to the test suites in both the incentive product and the open-source.

**Comment by speleo3** [comment]
CI with [pymol-testing](https://github.com/schrodinger/pymol-testing) implemented in #106
